[FLINK-14095][doc] Add Pravega connector in Flink Document (#1)
crazyzhou/flink: docs/dev/connectors/pravega.md

---
title: "Pravega Connector"
nav-title: Pravega
nav-parent_id: connectors
nav-pos: 6
---
<!--
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
-->

* This will be replaced by the TOC
{:toc}

## Pravega Introduction
[Pravega](http://pravega.io/) is an open-source, high-performance, durable, elastic and strongly-consistent
stream storage for continuous and unbounded data.

> Here are some need-to-know Pravega terminologies.

Reviewer: I suggest to replace this terminology list with the following paragraphs:

```
Pravega builds on the concept of *Stream*. A stream comprises one or more sequences of bytes. 
The unit of a stream is the *segment*. A *segment* is an unbounded sequence of bytes. A stream
can have one or more segments at any time, and the set of segments is allowed to change
dynamically with auto-scaling. Streams are grouped into scopes to avoid name collisions across
applications. 

One of the main APIs that Pravega provides is the event one. Events are serialized into bytes and
appended to segments. Pravega uses *routing keys* to map events to segments, they are
essentially sharded across open segments, and Pravega clients learn about open segments from
the control plane of Pravega.

Applications, including Flink jobs, read from a Pravega stream using reader groups. The readers
in a group share the workload of segments and coordinate the assignment of segments using a
Pravega feature that enables the coordination of state called *state synchronizer*. That coordination
is transparent to the application so that application does not have to handle dynamic changes to
the set of segments. This coordination also enables the generation of *checkpoints*, which are
essential to enable exactly-once semantics end-to-end with Flink. 

The foundation of checkpoints is an abstraction called *stream cut*: a collection of offsets spanning one or more segments. Checkpointing constitutes one of the main use cases of stream cuts, but stream cuts can additionally be used to set a position to begin reading or a position to end reading in a stream. To be able to resume or bound the computation over a stream of events, the Pravega API enables readers to obtain *stream cut* objects independent of checkpoints. 
```

Author: Replaced.

- **Stream**  A core abstraction -- `Stream` is provided to store unbounded events which aligns the design of Flink's DataStream API.
- **Scope**  Each `Stream` is under a namespace/tenant referred as `Scope`.
- **Segment**  Each `Stream` is split into a set of shards or partitions generally referred as `Segments` which can be auto-scaled.
- **Routing Key**  Every event written to a Pravega `Stream` has an associated `Routing Key` in an append-only way into a `Segment`.
A `Routing Key` is a string used by developers to group similar events which is also the basis for event ordering guarantee.
- **Streamcut**  A `StreamCut` represents a specific position in a Pravega `Stream`, which may be obtained from various API interactions with the Pravega client.

## Pravega connector

> Pravega connector can be used to build end-to-end real-time stream processing pipelines

Reviewer: We should say what the connector provides: source and sink implementations. As stated, it is too abstract. Here is a suggestion:

`The Pravega connector comprises source and sink implementations that enable jobs to use Pravega as an input of stream data and to store the output of jobs, respectively. The examples we show below in this document illustrate how an application sets it up to use the Pravega source and sink. See also the [Pravega Samples](https://github.com/pravega/pravega-samples) repository for more examples.`

Author: Replaced

(see [Samples](https://github.com/pravega/pravega-samples)) that use Pravega as the stream storage/message bus
and Apache Flink for computing engine over the streams.

### Features & Highlights

> - Exactly-once processing guarantees for both Reader and Writer, supporting end-to-end exactly-once processing pipelines

Reviewer: I'd remove the first part of this sentence and leave only: `Supports end-to-end exactly once semantics with checkpoints and transactions.`

Author: Replaced.

- Seamless integration with Flink's checkpoints and savepoints.
- Parallel Readers and Writers supporting high throughput and low latency processing.
- Table API support to access Pravega Streams for both Batch and Streaming use case.

### Dependency

> To use this connector, please pick the proper dependency and add it to your project:

Reviewer: Let's make explicit references to maven and gradlew, and adapt the samples below accordingly:

`To use the connector, add the appropriate dependency to your project via Maven or Gradlew`.

Author: I think as a Flink document, it should follow the Flink rule. Flink document only reference maven format for all its dependencies.
(e.g. https://ci.apache.org/projects/flink/flink-docs-stable/dev/connectors/kafka.html#usage 
https://ci.apache.org/projects/flink/flink-docs-release-1.9/getting-started/tutorials/datastream_api.html) 

There is not a template in the doc for switch in maven and other compiling tools, so I can only adopt your idea to keep showing these two together on the page. Your suggestion makes sense but I guess the Flink community wants to keep the tradition among the docs.

Reviewer 2: Should we update this section for the 0.7 release?

Reviewer 2: > There is not a template in the doc for switch in maven and other compiling tools, so I can only adopt your idea to keep showing these two together on the page. Your suggestion makes sense but I guess the Flink community wants to keep the tradition among the docs.

Here is a Flink quickstart document that shows both Maven and Gradle configs: https://ci.apache.org/projects/flink/flink-docs-stable/dev/projectsetup/java_api_quickstart.html#gradle

If it is just a short example, I think it is safe and helpful to include Gradle configuration stanzas here as well.

Author: Makes sense, I'll add the `gradle` section.

### Features & Highlights
- Supporting end-to-end exactly-once processing pipelines
- Seamless integration with Flink's checkpoints and savepoints.
- Parallel Readers and Writers supporting high throughput and low latency processing.
- Table API support to access Pravega Streams for both Batch and Streaming use case.

### Dependency
To use this connector, please pick the proper dependency and add it to your project:

{% highlight xml %}
<!-- Before Pravega 0.6 -->
<dependency>
  <groupId>io.pravega</groupId>
  <artifactId>pravega-connectors-flink_2.12</artifactId>
  <version>0.5.1</version>
</dependency>

<!-- After Pravega 0.6 -->
<dependency>
  <groupId>io.pravega</groupId>
  <artifactId>pravega-connectors-flink-1.9_2.12</artifactId>
  <version>0.6.1</version>
</dependency>
{% endhighlight %}

{% info %} As in the above example,

`1.9` is the Flink Major-Minor version which should be provided to find the right artifact.

`2.12` is the Scala version, now only 2.12 is supported.

`0.6.1` is the Pravega version, you can find the latest release on
the [GitHub Releases page](https://github.com/pravega/flink-connectors/releases).

> Since Pravega 0.6.0, the connector support multiple Flink versions.

Reviewer 2: ```suggestion
Since Pravega 0.6.0, the connector supports multiple Flink versions.
```

Author: Fixed

> Pravega connector maintains the compatibility for recent *three* minor version of Flink.

Reviewer 2: ```suggestion
Pravega connector maintains compatibility for the *three* most recent minor versions of Flink.
```

Author: Fixed

> See below table for the supported versions.

Reviewer 2: ```suggestion
See the table below for supported versions.
```

Author: Fixed
